### PR TITLE
feat(training): support Gemma 4 reserved-token SFT

### PR DESCRIPTION
**Pull request by lequytra**
## Description

Enable full-parameter Gemma 4 SFT to train customer control tokens by reusing the model's existing `<unusedN>` vocabulary rows instead of resizing the tokenizer or embedding matrix.

- Adds the persisted `fireworks_gemma4_reserved_token_map` tokenizer contract.
- Renames/activates selected reserved rows while asserting fixed tokenizer size and atomic encoding.
- Applies the same mapping in the main renderer and every DataLoader worker.
- Rejects the mapping for LoRA training because LoRA does not update base embedding rows.
- Adds a preparation helper and an end-to-end customer guide, including custom initialization and decode caveats.

Companion inference PR: [fw-ai/fireworks#37406](https://github.com/fw-ai/fireworks/pull/37406).

## Reviewer Ask

**Manual repro:** Run `/home/tran_le/cookbook/.venv/bin/python -m pytest training/tests/unit/test_tokenizers.py training/tests/unit/test_sft_loop.py -q`; expected result is `45 passed`.

**Review focus:** Please review the fixed-vocabulary invariant, the `alias -> <unusedN>` config shape, and whether the tokenizer metadata persistence matches custom-base promotion expectations.

**Not asking for:** Model embedding resize support or LoRA embedding training; this deliberately reuses existing Gemma 4 rows and requires `lora_rank=0`.

## Architecture / Code Overview

```mermaid
flowchart LR
    Config["alias → Gemma 4 <unusedN> map"] --> Load["Tokenizer load"]
    Load --> Validate["Validate BPE row + fixed vocab size"]
    Validate --> Atomic["Register atomic alias at existing ID"]
    Atomic --> Workers["Main renderer + DataLoader workers"]
    Workers --> SFT["Full-parameter SFT"]
    SFT --> Embedding["Selected embedding/output rows learn"]
    Atomic --> Saved["tokenizer_config.json persists mapping"]
    Saved --> Inference["Companion inference loader"]
```

## Testing

- [x] Added/updated tests
- [x] Tested manually

- Unit tests: `45 passed`.
- Real Gemma 4 tokenizer: vocabulary stayed at `262144`; aliases resolved atomically to existing IDs `256035` and `256036`.
- Real Gemma 4 renderer: the aliased token appeared in the supervised assistant span with loss weight `1.0`.
- Formatting: focused Black check and `git diff --check`.

## Performance

The mapping is applied once per tokenizer/render-worker initialization. It adds no per-example work after startup and does not change GPU kernels, synchronization, or the training step.

## Surface Consistency

- [x] Training CLI, tokenizer persistence, renderer workers, docs, and companion inference support use the same config key and mapping semantics.

## Checklist

- [x] Agent-reviewed and self-reviewed the diff
- [x] Minimum necessary diff
- [x] Tests and documentation added
- [x] No secrets or credentials in the diff
- [x] Visual diagram included

### CODE REVIEW

**Comment by lequytra** [comment]
Closing this direct mirror PR: cookbook changes must be reviewed from `fw-ai/fireworks/public-repos/cookbook` and promoted by the repository workflow. The identical training changes are now included with inference in https://github.com/fw-ai/fireworks/pull/37406.
